File: apps/web/src/lib/d1.ts
import "server-only";

/**
 * Cloudflare D1 over the REST API — the TypeScript twin of the worker's `d1.py`.
 *
 * The web app deploys on Node/Vercel, not the Workers runtime, so it talks to D1
 * over HTTPS with the same account/token/database env the worker uses. Every call
 * funnels through `execute()`; when D1 is unconfigured it returns `[]` so
 * `next build` and local dev render an empty (not crashing) UI.
 */

const API_BASE = "https://api.cloudflare.com/client/v4";

function config() {
  const accountId = process.env.CLOUDFLARE_ACCOUNT_ID;
  const token = process.env.CLOUDFLARE_API_TOKEN;
  const databaseId = process.env.D1_DATABASE_ID;
  if (!accountId || !token || !databaseId) return null;
  return { accountId, token, databaseId };
}

export function isConfigured(): boolean {
  return config() !== null;
}

type Primitive = string | number | null;

export async function execute<T = Record<string, unknown>>(
  sql: string,
  params: Primitive[] = [],
): Promise<T[]> {
  const cfg = config();
  if (!cfg) return [];

  const res = await fetch(
    `${API_BASE}/accounts/${cfg.accountId}/d1/database/${cfg.databaseId}/query`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${cfg.token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ sql, params }),
      // D1 data changes per pipeline run; never serve stale cached reads.
      cache: "no-store",
    },
  );

  if (!res.ok) {
    throw new Error(`D1 HTTP ${res.status}: ${await res.text()}`);
  }
  const body = (await res.json()) as {
    success: boolean;
    errors?: unknown;
    result?: Array<{ results?: T[] }>;
  };
  if (!body.success) {
    throw new Error(`D1 query failed: ${JSON.stringify(body.errors)}`);
  }
  return body.result?.[0]?.results ?? [];
}

/** Run a write and return the first RETURNING row (or null). */
export async function executeOne<T = Record<string, unknown>>(
  sql: string,
  params: Primitive[] = [],
): Promise<T | null> {
  const rows = await execute<T>(sql, params);
  return rows[0] ?? null;
}

File: apps/worker/d1.py
"""Low-level Cloudflare D1 client (REST API).

The worker runs on Railway (Python), so it talks to D1 over HTTP rather than a
Workers binding:

    POST https://api.cloudflare.com/client/v4/accounts/{acct}/d1/database/{db}/query
    Authorization: Bearer {CLOUDFLARE_API_TOKEN}
    { "sql": "...", "params": [...] }

Everything funnels through `execute()` / `_execute()` so the repository layer in
`db.py` stays pure SQL-building and the tests can monkeypatch a single seam. If
D1 is not configured, calls raise `D1NotConfigured`; callers in `db.py` catch
this and degrade to no-ops so the offline pipeline/tests keep working.
"""
from __future__ import annotations

import logging
from typing import Any

import httpx

from config import settings

logger = logging.getLogger(__name__)

_API_BASE = "https://api.cloudflare.com/client/v4"


class D1NotConfigured(RuntimeError):
    pass


class D1Error(RuntimeError):
    pass


def is_configured() -> bool:
    return bool(
        settings.cloudflare_account_id
        and settings.cloudflare_api_token
        and settings.d1_database_id
    )


def _execute(sql: str, params: list[Any] | None = None) -> list[dict]:
    """Run one SQL statement against D1 and return its result rows.

    Single seam for the HTTP call — tests monkeypatch this. D1 binds params
    positionally with `?` placeholders; values must be primitives (str/int/
    float/None), so callers JSON-encode arrays/objects before binding.
    """
    if not is_configured():
        raise D1NotConfigured("Cloudflare D1 is not configured (see .env.example)")
    url = (
        f"{_API_BASE}/accounts/{settings.cloudflare_account_id}"
        f"/d1/database/{settings.d1_database_id}/query"
    )
    resp = httpx.post(
        url,
        headers={"Authorization": f"Bearer {settings.cloudflare_api_token}"},
        json={"sql": sql, "params": params or []},
        timeout=30,
    )
    resp.raise_for_status()
    body = resp.json()
    if not body.get("success", False):
        raise D1Error(f"D1 query failed: {body.get('errors')}")
    # The query endpoint returns a list of result objects (one per statement).
    results = body.get("result", [])
    if results and isinstance(results, list):
        return results[0].get("results", []) or []
    return []


def execute(sql: str, params: list[Any] | None = None) -> list[dict]:
    """Public wrapper with logging; returns [] on a not-configured environment."""
    try:
        return _execute(sql, params)
    except D1NotConfigured:
        logger.debug("D1 not configured; skipping query: %s", sql.split("\n", 1)[0])
        return []
    except httpx.HTTPError as exc:
        raise D1Error(str(exc)) from exc
